string.c

```diff
@@ -5732,7 +5732,7 @@ rb_str_setbyte(VALUE str, VALUE index, VALUE value)
 		ENC_CODERANGE_SET(str, ENC_CODERANGE_VALID);
 	    goto end;
 	  case ENC_CODERANGE_VALID:
-	    left = rb_enc_left_char_head(head, ptr, head+len, enc);
+            left = rb_enc_left_char_head(head, (char *)ptr, head+len, enc);
```

Comment:
Shouldn't `ptr` be a plain `char` pointer?

Reply:
That works _for me_, but could break this line in some environment:

https://github.com/ruby/ruby/blob/3978a2abb86ae85325bdc8fedb9bf08ed0d65248/string.c#L5737

If `ptr` is plain `char`, and if the environment chooses it's signed, then unsigned to signed conversion is allowed to fail.

Reply:
It may need a cast, but I think that casting signedness is preferable to pointer punning.

Reply:
Well, this case is not "punning", still I prefer value-casts to pointer-casts.